### ISSUE TITLE
Problem while using guzzle with pthreads

### ISSUE THREAD

**Issue by Massimiliano87** [guzzle/guzzle#1398]
Hi, 
I've found a pretty odd problem while using guzzle 6 inside threads. Here is the full explanation:
My script consists of 2 parts; the first calls an external API, fetches some data and store it inside a Redis cache. The second part uses that data to call other APIs by opening a pool of threads, which uses guzzle to call those APIs.
The problem is that IF in the first part I don't have the data in cache (and so i have to make a call to the first API) in the second part guzzle returns a catchable fatal error: "Argument 1 passed to GuzzleHttp\Promise\Promise::__construct() must be callable, array given, called in /var/www/hs-pv4/vendor/guzzlehttp/promises/src/FulfilledPromise.php on line 34 and defined in /var/www/hs-pv4/vendor/guzzlehttp/promises/src/Promise.php on line 23". 
Further investigations led to the discovery that ONLY if i made the first call to my API in the first part of my script the object returned by the guzzle "queue" function (guzzlehttp/promises/src/functions.php) was an instance of the class CurlMultiExec instead of being an instance of the class TaskQueue.
Instead if in the first part i already had the data in my cache (and so i didn't have to call my API) the second part worked without any problem.

Has anyone experienced this problem?

**Comment by geki-yaba** [comment]
Well, I have a similar problem using Guzzle in worker threads with composer and its autoloading. If I do not create at least one Instance of the Classes I need in the worker threads in the main thread first, I get "Class not found" errors.

That workaround is okay for me.

Another issue I found is, that serialization of GuzzleHttp/Client fails for a Closure. That can be worked around by implementing __sleep() to release the client instance and __wakeup() to recreate it in the Class which owns GuzzleHttp/Client.

Things one may face with threading.

**Comment by mtdowling** [comment]
I would guess that the underlying cURL handles, and PHP's integration with cURL, have an issue being shared across multiple threads. You may have to create unique clients for each thread.

> Another issue I found is, that serialization of GuzzleHttp/Client fails for a Closure. That can be worked around by implementing __sleep() to release the client instance and __wakeup() to recreate it in the Class which owns GuzzleHttp/Client.

Yeah that's true, Guzzle client's cannot be serialized due to heavy use of closures.

**Comment by Massimiliano87** [comment]
> I would guess that the underlying cURL handles, and PHP's integration with cURL, have an issue being shared across multiple threads. You may have to create unique clients for each thread.

The fact is that in each Thread I instantiate new Client and Request objects, i do not use a single instance of the Client class across all of my threads.

**Comment by mtdowling** [comment]
Ah, I see. The promise queue is a shared, static resource that all promises use. We'd probably need some way to have a per/thread promise queue in order to work around this.

**Comment by Massimiliano87** [comment]
Yes, i think that would be a solution. As for now it's nice to have had a confirmation about the core of the problem, at least now I know for certain what was causing the malfunctioning inside my project.

**Comment by mtdowling** [comment]
I don't think Guzzle will or should be refactored to support pthreads. Guzzle's concurrency story is single-threaded nonblocking IO rather than threading. Perhaps this could be addressed if Guzzle 7 were ever released, but I have no plans for that any time soon.

**Comment by geki-yaba** [comment]
Just in case someone is interested:

WRT threading, one may autoload guzzle in each worker thread, see issue #1434. Works just fine.
WRT closures, either implement a GuzzleHttp/Client variant with SuperClosures[0][1] or wait for PHP bug [#67569](https://bugs.php.net/bug.php?id=67569).

[0] https://github.com/jeremeamia/super_closure
[1] https://github.com/opis/closure

**Comment by AleksSv** [comment]
geki-yaba, just to clarify,  what do you mean exactly by "Works just fine". Can you get GuzzleHttp to work in parallel just with some workarounds? For example if I have 4 threads that need to do 4 separate Guzzlehttp requests, your workarounds will allow them to all be processed simultaneously?

**Comment by geki-yaba** [comment]
I do thousands of requests threaded, yes. A partial example of worker thread doing the magic:
Finally, do create guzzle http client exclusively in that Thread.

class Thread extends Worker
{   
    public function run()
    {   # do composer autoloading in threads
        require dirname(__DIR__) . '/vendor/autoload.php';

        ...
    }
}

**Comment by AleksSv** [comment]
I appreciate the quick response geki-yaba.
I use a lot of different composer libraries so I was having trouble doing what you did exactly.
However, what I found is that I could download a standalone directory of Guzzlehttp ( i also needed fabpot/Goutee) from 

https://php-download.com/package/guzzlehttp/guzzle
https://php-download.com/package/fabpot/goutte

So for all my classes, I do 
`require_once __DIR__ . "/vendor/autoload.php";
`
While inside the thread, I do

```
 public function run(){
        require_once __DIR__ . '/guzzlehttp_guzzle_6.3.3.0_require/vendor/autoload.php';
        require_once __DIR__ . '/fabpot_goutte_3.1.2.0_require/vendor/autoload.php';


}
```

With those requires, it appears to "work just fine" as you said. 
Now my Guzzlehttp processing is being done with 4 threads working in parallel popping off jobs from a shared array, instead of all sequentially under 1 thread like it was before.

https://github.com/krakjoe/pthreads/blob/master/examples/StackableArray.php

Are there any race conditions we need to worry about?
Your solution is so straightforward and effective, I'm kinda paranoid there's something I'm missing.

**Comment by geki-yaba** [comment]
All my composer libraries are only used in the threads. The main thread does initialization, signal setup and setup of my global "context" instance only and then uses a Threaded instance to pass integer indices to threads. The Worker instances are constructed with a reference to my global "context" instance and query that "context" instance for data wrt the passed integer indice from within their threads.

I have no race conditions, i.e. Logging(generic: pass data to global "context" instance) needs serialization. My DB access needs no serialization luckily. Check your storage backend. Normal threading business.

It is all quite straight-forward. How to implement is an exercise for the reader.
